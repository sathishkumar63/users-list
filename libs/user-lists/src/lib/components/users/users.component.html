<div class="users">
  <div class="control-panel">
    <form>
      <label class="select"
        >Select column to filter:
        <select
          name="filterProperty"
          id="filterProperty"
          [(ngModel)]="filterProperty"
        >
          <option *ngFor="let opt of options" [value]="opt.value">
            {{ opt.text }}
          </option>
        </select>
      </label>
      <label class="input"
        >Enter filter keywords:
        <input
          type="text"
          placeholder="filter keywords go here..."
          name="filterValue"
          id="filterValue"
          [(ngModel)]="filterValue"
        />
      </label>
    </form>
  </div>
  <div class="users-list-wrapper">
    <table class="users-list">
      <thead>
        <tr>
          <th scope="col">Name</th>
          <th scope="col">User Name</th>
          <th scope="col">E-mail</th>
          <th scope="col">Phone</th>
          <th scope="col">Website</th>
        </tr>
      </thead>
      <tbody>
        <!-- Filtering User Data Using Pipes -->
        <tr
          *ngFor="
            let _userList of userList | Filter: filterProperty:filterValue
          "
        >
          <td>{{ _userList.name }}</td>
          <td>{{ _userList.username }}</td>
          <td>{{ _userList.email }}</td>
          <td>{{ _userList.phone }}</td>
          <td>{{ _userList.website }}</td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
