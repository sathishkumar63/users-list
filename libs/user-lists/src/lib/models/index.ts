export * from './user.interface';
export * from './users.mock.data';
