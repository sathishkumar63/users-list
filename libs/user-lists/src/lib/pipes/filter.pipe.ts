import { Pipe, PipeTransform, Injectable } from '@angular/core';

@Pipe({
  name: 'Filter',
})
@Injectable()
export class FilterPipe implements PipeTransform {
  /**
   * Transform
   *
   * @param {any[]} items
   * @param {string} searchText
   * @param {string} property
   * @returns {any}
   */
  transform(items: any[], property: string, searchText: string): any[] {
    if (!items) return [];
    if (!searchText || searchText.length == 0) return items;
    return items.filter(
      (obj) =>
        obj[property]?.toLowerCase().indexOf(searchText.toLowerCase()) != -1
    );
  }
}
