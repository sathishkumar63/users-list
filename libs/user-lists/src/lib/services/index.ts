export * from './users/users.service';
export * from './users/users.mock.service';
