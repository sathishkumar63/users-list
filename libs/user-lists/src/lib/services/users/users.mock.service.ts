import { Observable, of } from 'rxjs';
export class MockUsersService {
  getJSON(): Observable<any> {
    return of();
  }
}
