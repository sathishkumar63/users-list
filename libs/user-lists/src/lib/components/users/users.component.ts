import { Component, OnInit } from '@angular/core';
import { FilterOption } from '../../models';
import { UsersService } from '../../services';

@Component({
  selector: 'users-list',
  templateUrl: './users.component.html',
  styleUrls: ['./users.component.scss'],
})
export class UsersComponent implements OnInit {
  userList: any = [];
  filterProperty = 'name';
  filterValue = '';
  options: FilterOption[] = [
    {
      value: 'name',
      text: 'Name',
    },
    {
      value: 'username',
      text: 'User Name',
    },
    {
      value: 'email',
      text: 'Email',
    },
    {
      value: 'phone',
      text: 'Phone',
    },
    {
      value: 'website',
      text: 'Website',
    },
  ];

  constructor(private userService: UsersService) {}

  ngOnInit() {
    /** `subscribe` actually initiates the call to the server **/
    this.userService.getJSON().subscribe((data) => {
      this.userList = data;
    });
  }
}
