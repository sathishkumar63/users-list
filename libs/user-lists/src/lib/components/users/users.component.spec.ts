import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { of } from 'rxjs';
import { mockUsers } from '../../models';
import { FilterPipe } from '../../pipes';
import { MockUsersService, UsersService } from '../../services';

import { UsersComponent } from './users.component';

describe('UsersComponent', () => {
  let component: UsersComponent;
  let fixture: ComponentFixture<UsersComponent>;
  let service: UsersService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [UsersComponent, FilterPipe],
      imports: [ReactiveFormsModule, FormsModule],
      providers: [
        {
          provide: UsersService,
          useClass: MockUsersService,
        },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(UsersComponent);
    service = TestBed.inject(UsersService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should create snapshot', () => {
    component.userList = mockUsers;
    fixture.detectChanges();
    const debugElement: HTMLElement = fixture.nativeElement;
    expect(debugElement).toMatchSnapshot();
  });

  describe('ngOnInit', () => {
    test('should make expected call', () => {
      // arrange
      jest.spyOn(service, 'getJSON');
      // act
      component.ngOnInit();
      // assert
      expect(service.getJSON).toBeCalled();
    });

    test(
      'should get users when calling getJSON()',
      waitForAsync(() => {
        // arrange
        jest.spyOn(service, 'getJSON').mockReturnValue(of(mockUsers));
        // act
        service.getJSON().subscribe((data) => {
          component.userList = data;
        });
        // assert
        expect(component.userList).toEqual(mockUsers);
      })
    );
  });
});
