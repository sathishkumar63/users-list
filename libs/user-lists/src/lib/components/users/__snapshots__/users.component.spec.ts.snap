// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersComponent should create snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="users"
  >
    <div
      class="control-panel"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <label
          class="select"
        >
          Select column to filter: 
          <select
            class="ng-untouched ng-pristine ng-valid"
            id="filterProperty"
            name="filterProperty"
          >
            <option
              value="name"
            >
               Name 
            </option>
            <option
              value="username"
            >
               User Name 
            </option>
            <option
              value="email"
            >
               Email 
            </option>
            <option
              value="phone"
            >
               Phone 
            </option>
            <option
              value="website"
            >
               Website 
            </option>
            
          </select>
        </label>
        <label
          class="input"
        >
          Enter filter keywords: 
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="filterValue"
            name="filterValue"
            placeholder="filter keywords go here..."
            type="text"
          />
        </label>
      </form>
    </div>
    <div
      class="users-list-wrapper"
    >
      <table
        class="users-list"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Name
            </th>
            <th
              scope="col"
            >
              User Name
            </th>
            <th
              scope="col"
            >
              E-mail
            </th>
            <th
              scope="col"
            >
              Phone
            </th>
            <th
              scope="col"
            >
              Website
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Leanne Graham
            </td>
            <td>
              Bret
            </td>
            <td>
              Sincere@april.biz
            </td>
            <td>
              1-770-736-8031 x56442
            </td>
            <td>
              hildegard.org
            </td>
          </tr>
          <tr>
            <td>
              Ervin Howell
            </td>
            <td>
              Antonette
            </td>
            <td>
              Shanna@melissa.tv
            </td>
            <td>
              010-692-6593 x09125
            </td>
            <td>
              anastasia.net
            </td>
          </tr>
          
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
